src/pal/src/sharedmemory/sharedmemory.cpp

```diff
@@ -0,0 +1,1114 @@
+{
+    _ASSERTE(m_name != nullptr);
+    return m_name;
+}
+
+SIZE_T SharedMemoryId::GetNameCharCount() const
+{
+    _ASSERTE(m_name != nullptr);
+    return m_nameCharCount;
+}
+
+bool SharedMemoryId::IsSessionScope() const
+{
+    _ASSERTE(m_name != nullptr);
+    return m_isSessionScope;
+}
+
+bool SharedMemoryId::Equals(SharedMemoryId *other) const
+{
+    return
+        GetNameCharCount() == other->GetNameCharCount() &&
+        IsSessionScope() == other->IsSessionScope() &&
+        strcmp(GetName(), other->GetName()) == 0;
+}
+
+SIZE_T SharedMemoryId::AppendSessionDirectoryName(
+    char (&path)[SHARED_MEMORY_MAX_FILE_PATH_CHAR_COUNT + 1],
+    SIZE_T pathCharCount) const
+{
+    if (IsSessionScope())
+    {
+        pathCharCount = SharedMemoryHelpers::CopyString(path, pathCharCount, SHARED_MEMORY_SESSION_DIRECTORY_NAME_PREFIX);
+        int sessionIdCharCount =
+            sprintf_s(
+                &path[pathCharCount],
+                _countof(path) - pathCharCount,
+                "%u",
+                GetCurrentSessionId());
+        _ASSERTE(sessionIdCharCount > 0);
+        pathCharCount += sessionIdCharCount;
+    }
+    else
+    {
+        pathCharCount = SharedMemoryHelpers::CopyString(path, pathCharCount, SHARED_MEMORY_GLOBAL_DIRECTORY_NAME);
+    }
+
+    _ASSERTE(pathCharCount <= SHARED_MEMORY_MAX_FILE_PATH_CHAR_COUNT);
+    return pathCharCount;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// SharedMemorySharedDataHeader
+
+SIZE_T SharedMemorySharedDataHeader::DetermineTotalByteCount(SIZE_T dataByteCount)
+{
+    return SharedMemoryHelpers::AlignUp(sizeof(SharedMemorySharedDataHeader) + dataByteCount, PAGE_SIZE);
+}
+
+SharedMemorySharedDataHeader::SharedMemorySharedDataHeader(SharedMemoryType type, UINT8 version)
+    : m_type(type), m_version(version)
+{
+}
+
+SharedMemoryType SharedMemorySharedDataHeader::GetType() const
+{
+    return m_type;
+}
+
+UINT8 SharedMemorySharedDataHeader::GetVersion() const
+{
+    return m_version;
+}
+
+void *SharedMemorySharedDataHeader::GetData()
+{
+    return this + 1;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// SharedMemoryProcessDataHeader
+
+SharedMemoryProcessDataHeader *SharedMemoryProcessDataHeader::CreateOrOpen(
+    LPCSTR name,
+    SharedMemorySharedDataHeader requiredSharedDataHeader,
+    SIZE_T sharedDataByteCount,
+    bool createIfNotExist,
+    bool *createdRef)
+{
+    _ASSERTE(name != nullptr);
+    _ASSERTE(sharedDataByteCount != 0);
+    _ASSERTE(!createIfNotExist || createdRef != nullptr);
+    _ASSERTE(SharedMemoryManager::IsCreationDeletionProcessLockAcquired());
+    _ASSERTE(!SharedMemoryManager::IsCreationDeletionFileLockAcquired());
+
+    if (createdRef != nullptr)
+    {
+        *createdRef = false;
+    }
+
+    SharedMemoryId id(name);
+
+    struct AutoCleanup
```

Comment:
In some cases, things need to be cleaned up in a specific order, some things need to be skipped and others not, depending on where we are, etc. With holders, that cleanup logic would be complicated to understand. With this method, it is made directly clear, and is flexible for doing any sort of cleanup. I'll take a look and will reconsider later.